optimization/sggp/sggp-algorithm-with-test.metta

```diff
@@ -0,0 +1,597 @@
+        $expression
+        (empty)
+    )
+)
+;;The following function leverages Non-determinism to filter a function
+;; based on a predicated
+(= (filterList $compSign $func $expressions $threshold)
+    (let $expressions' (superpose $expressions)
+        (collapse (filter' $compSign $func $expressions' $threshold))
+    )
+)
+;;This procedure implements the reinforcing the production rule that is available.
+(= (reinforceProd ($lhs $rhs $weight) $learnRate $cardinality $stocSpace)
+    (let ($parent $child $updatedWeight) (updateWeight ($lhs $rhs $weight) $learnRate $cardinality inc)
+        (update-atom $stocSpace ($lhs $rhs $weight) ($parent $child $updatedWeight))
+    )
+)
+;; This procedure works if multiple weights happen 
+(= (reinforceProd ($prod1 $prod2) $learnRate $cardinality $stocSpace)
+    (let* (
+        ($res (reinforceProd $prod1 $learnRate $cardinality $stocSpace))
+        ($res' (reinforceProd $prod2 $learnRate $cardinality $stocSpace))
+        (() (println! (reinforceProd works)))
+    )
+        ()
+    )
+)
+;; This procedure implements the punishing a specific production rule that is available.
+(= (punishProd ($lhs $rhs $weight) $learnRate $cardinality $stocSpace)
+    (let ($parent $child $updatedWeight) (updateWeight ($lhs $rhs $weight) $learnRate $cardinality dec)
+        (update-atom $stocSpace ($lhs $rhs $weight) ($parent $child $updatedWeight))
+    );; (exploitInstances $rhs $grammarSpace $stocSpace $explorationCoeff (- $depth 1))
+)
+(= (punishProd ($prod1 $prod2) $learnRate $cardinality $stocSpace)
+    (let* (
+        ($res (punsihProd $prod1 $learnRate $cardinality $stocSpace))
+        ($res' (punishProd $prod2 $learnRate $cardinality $stocSpace))
+    )
+        ()
+    )
+)
+
+;; This procedure implements the reinforcement of the reinforcement of a grammar given a Tree
+(= (reinforceGrammarbasedOnTree $tree $learnRate $cardinality $grammarSpace $stocSpace)
+    (let* (
+        ($relations (getRelations $tree ()))
+        ($prodRules (collapse (getRuleFromSubTree (superpose $relations) $stocSpace)))
+        ;; (() (println! (production rules $prodRules)))
+        ;; ($reinforcedOutput (reinforceTreeHelper $prodRules $learnRate $cardinality $stocSpace))
+        
+    )
+        (reinforceTreeHelper $prodRules $learnRate $cardinality $stocSpace)
+        
+        ;;(collapse (reinforceProd (superpose $prodRules) $learnRate $cardinality $stocSpace))
+        ;; $reinforcedOutput
+    )
+)
+
+;; The below piece of code punishes the production rule in a stochastic grammar based on a tree
+(= (punishGrammarbasedOnTree $tree $learnRate $cardinality $grammarSpace $stocSpace)
+    (let* (
+        ($relations (getRelations $tree ()))
+        ($prodRules (collapse (getRuleFromSubTree (superpose $relations) $stocSpace)))
+        ;; (println (production rules $prodRules))
+    )
+        ;;(collapse (punishProd (superpose $prodRules) $learnRate $cardinality $stocSpace))
```

Comment:
Why is this code commented out? Is there a case where this commented out code would be usefull? If so, add that case as a comment so that it won't be forgotten in the future. If it won't be useful remove it.